### PR TITLE
Refactor: move RowGroupPredicateBuilder into its own module, rename to PruningPredicateBuilder

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Part of https://github.com/apache/arrow-datafusion/issues/363

 # Rationale for this change

As explained on https://github.com/apache/arrow-datafusion/issues/363 the high level idea goal is to make the parquet row group pruning logic generic to any types of min/max statistics (not just parquet metadata)

# What changes are included in this PR?
1. Move code out of parquet.rs and into its own module
2. Rename to StatsPredicateBuilder to reflect its purpose

No changes in functionality are intended

The PR contains two commits to help reviewers

# Are there any user-facing changes?
The `pub struct RowGroupPredicateBuilder` struct was renamed

### CODE REVIEW

**Comment by alamb** [comment]
FYI @yordan-pavlov and @returnString

**Review by jorgecarleitao — APPROVED** [review]
I have not read it in detail but I assume it is just a code movement and rename and therefore 👍

File: datafusion/src/physical_optimizer/pruning.rs [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+    /// This will translate the filter expression into a statistics predicate expression
+    ///
+    /// For example,  `(column / 2) = 4` becomes `(column_min / 2) <= 4 && 4 <= (column_max / 2))`
+    pub fn try_new(expr: &Expr, schema: Schema) -> Result<Self> {
+        // build predicate expression once
+        let mut stat_column_req = Vec::<(String, StatisticsType, Field)>::new();
+        let logical_predicate_expr =
+            build_predicate_expression(expr, &schema, &mut stat_column_req)?;
+        // println!(
+        //     "PruningPredicateBuilder::try_new, logical_predicate_expr: {:?}",
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/pruning.rs:71]: Those println comments could be removed?

Comment (Dandandan) [reply datafusion/src/physical_optimizer/pruning.rs:71]: (they were there in the original code but I think would be better to remove them)

File: datafusion/src/physical_optimizer/pruning.rs [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+
+    // fn column_expr(&self) -> &Expr {
+    //     self.column_expr
+    // }
+
+    fn scalar_expr(&self) -> &Expr {
+        self.scalar_expr
+    }
+
+    // fn column_name(&self) -> &String {
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/pruning.rs:259]: Could be removed

File: datafusion/src/physical_optimizer/pruning.rs [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+        match op {
+            Operator::Lt => Operator::Gt,
+            Operator::Gt => Operator::Lt,
+            Operator::LtEq => Operator::GtEq,
+            Operator::GtEq => Operator::LtEq,
+            _ => op,
+        }
+    }
+
+    // fn column_expr(&self) -> &Expr {
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/pruning.rs:251]: Could be removed

File: datafusion/src/physical_optimizer/pruning.rs [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+    }
+
+    // fn column_name(&self) -> &String {
+    //     &self.column_name
+    // }
+
+    fn is_stat_column_missing(&self, statistics_type: StatisticsType) -> bool {
+        self.stat_column_req
+            .iter()
+            .filter(|(c, t, _f)| c == &self.column_name && t == &statistics_type)
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/pruning.rs:266]: This code could use `.any` instead of filter + count.

**Comment by Dandandan** [comment]
I also think this is looking good. I think we might clean things up a bit while touching the code, I added some suggestions.

File: datafusion/src/physical_optimizer/pruning.rs (hunk 3 above)
Comment (alamb) [reply datafusion/src/physical_optimizer/pruning.rs]: in 688dde41b748801183d3484c6116442c22bdf591

File: datafusion/src/physical_optimizer/pruning.rs (hunk 2 above)
Comment (alamb) [reply datafusion/src/physical_optimizer/pruning.rs]: in 688dde41b748801183d3484c6116442c22bdf591

**Comment by alamb** [comment]
> I also think this is looking good. I think we might clean things up a bit while touching the code, I added some suggestions.

Cleaned per @Dandandan 's suggestions

**Review by Dandandan — APPROVED** [review]
I like it, great use of the physical optimizer

**Pull request merged**